UI: Add toggle switch type when rendering checkbox

lloyd-teamtangible opened the pull request:
Hi, @eliot-akira!

I've updated the UI for the plugin features, and I’d love to hear your thoughts on the changes if you have some time to check it out. Thanks!

![image](https://github.com/user-attachments/assets/2a0ebdac-4a3a-40ee-bbba-2e9a16808642)

eliot-akira commented:
That's lovely! I'll some inline comments for feedback.

eliot-akira commented on plugin/settings/checkbox.php:
```diff
@@ -13,28 +13,52 @@
+    'label',
+    'description'
   ] as $key) {
     $$key = $config[$key];
```
```suggestion
    $$key = $config[$key] ?? '';
```

This change will prevent "undefined property" warning for the newly added `description`, which existing uses of the function haven't defined. It makes it optional.

eliot-akira commented on plugin/settings/settings.css:
```diff
@@ -311,3 +309,87 @@
+  font-size: .9rem;
+}
+
+.tangible-slider {
```
I would name this CSS class differently, `slider` is a bit ambiguous. Maybe `switch-slider`?

There's a name for this specific UI element, but I don't recall. I think I saw it in the [Open UI project](https://open-ui.org).

> The purpose of the Open UI, a [W3C Community Group](https://www.w3.org/community/open-ui/), is to allow web developers to style and extend built-in web UI components and controls, such as <select> dropdowns, checkboxes, radio buttons, and date/color pickers.
>
> To do that, we’ll need to fully specify the **component parts, states, and behaviors** of the built-in controls, as well as necessary **accessibility** requirements.

Here's the Switch component.

- https://open-ui.org/components/switch/

Perfect, it's a good overview, including:

#### Anatonmy

- checkedContent
- uncheckedContent
- root
- thumb
- loadingContent
- track

There's no description of each element. Looks like "track" is the word we want.

eliot-akira commented on plugin/settings/checkbox.php:
```diff
@@ -13,28 +13,52 @@
  * they use a smart JavaScript function to gather form fields data.
  */
-function render_setting_field_checkbox($config) {
+function render_setting_field_checkbox($config, $type='checkbox') {
```
Could add a comment in the docblock above, what values are accepted for new this function argument?

eliot-akira replied:
Actually, I think I'd prefer this to be a `type` property of the `$config` argument.

eliot-akira commented on plugin/settings/checkbox.php:
```diff
@@ -13,28 +13,52 @@
+function render_setting_field_checkbox($config, $type='checkbox') {
 
   foreach ([
     'name',
```
```suggestion
    'type',
    'name',
```

eliot-akira commented on plugin/settings/checkbox.php:
```diff
@@ -13,28 +13,52 @@
-    <?php echo esc_html($label); ?>
-  </label>
-  <?php
+  if ($type === 'checkbox'): ?>
```
```suggestion
  if (empty($type)): ?>
```

It's `checkbox` by default for backward compatibility.

eliot-akira commented:
There's a reference where I read about the Switch component in detail, in [ARIA Authoring Practices](https://www.w3.org/WAI/ARIA/apg/patterns/).

- [Switch Pattern](https://www.w3.org/WAI/ARIA/apg/patterns/switch/)
- [Switch Example Using HTML Checkbox Input](https://www.w3.org/WAI/ARIA/apg/patterns/switch/examples/switch-checkbox/)

The first link is about **keyboard navigation and ARIA states**. I'd like to come back to this topic later. The second link is like an ideal switch with well-structured HTML/CSS/JS. Here's the code from the docs site repo.

- https://github.com/w3c/aria-practices/tree/main/content/patterns/switch

In the folder, we can see the [HTML structure](https://github.com/w3c/aria-practices/blob/af8f4c1d97ead402022d704754c3d382c4a2cef2/content/patterns/switch/examples/switch-checkbox.html#L52-L62) of the component, the [CSS for switch](https://github.com/w3c/aria-practices/blob/main/content/patterns/switch/examples/css/switch.css) and [switch checkbox](https://github.com/w3c/aria-practices/blob/main/content/patterns/switch/examples/css/switch-checkbox.css).

Ideally, I'd like to refer to the above example for developing a **switch component for general-purpose use**. Then we can update this specific "checkbox and switch for plugin settings" to use a generic component from a shared library.

Speaking of generic components, I see there's anoter UI concept in this PR, the "card". I like how you created a generic "card" and a specific "feature card". That will make it easier if we want to move the "card" component into a library in the future.

GabrielGallagher commented:
@eliot-akira Are we sure we want to be loading in bootstrap or really grabbing anything from it at all in the design system? Maybe @juliacanzani has some thoughts.

eliot-akira commented:
The Design library is a complete rewrite of Bootstrap, so it can be made into whatever we need, removing unused files, etc. Using an existing library saves us a lot of time, as it comes with a good foundation of well-written code.

It's gone through several iterations, first as a utility class library like [BassCSS](https://basscss.com/) and [Tailwind](https://tailwindcss.com/); then as a class-less minimal framework like [PicoCSS](https://picocss.com/); in the end, I forked and took apart Bootstrap v5, rewriting it as a modular library with optional CSS prefix on all classes - for use in WordPress plugins, it must work well within any site theme and global stylesheet.

lloyd-teamtangible commented:
Hi, @eliot-akira!

Thanks a ton for taking the time to review my changes and for all the helpful feedback. I really appreciate it!

I just pushed a new commit addressing your comments. If I missed anything, just let me know!

Thanks again!

eliot-akira commented:
Wonderful, I appreciate the contribution and your patience. I tried it on a local site with a test plugin. Nice, I see you added keyboard navigation and ARIA states.

As you can see, things are in various states of development. I wanted to share my thoughts on the direction I'm hoping to take these related projects, the Framework, Fields, Design modules.

If you have any ideas for features or changes, please feel free to open an issue or join an existing one for discussion.
